doc/developer/design/20231127_decoupled_isolated_coordinator.md

```diff
@@ -0,0 +1,487 @@
+  and reading thereof. It is explicitly a pTVC that carries incremental changes
+  to the catalog!
+- TIMESTAMP ORACLE is a linearizable object that maintains read/write
+  timestamps for timelines.
+
+And we also change how some of the existing components work.
+
+### TIMESTAMP ORACLE
+
+The TIMESTAMP ORACLE can be considered a service that can be called from any
+component that needs it, and it provides linearizable timestamps and/or records
+writes as applied.
+
+- `ReadTs(timeline) -> Timestamp`: returns the current read timestamp.
+
+  This timestamp will be greater or equal to all prior values of
+  `ApplyWrite()`, and strictly less than all subsequent values of `WriteTs()`.
+
+- `WriteTs(timeline) -> Timestamp`: allocates and returns monotonically
+  increasing write timestamps.
+
+  This timestamp will be strictly greater than all prior values of `ReadTs()`
+  and `WriteTs`.
+
+- `ApplyWrite(timeline, timestamp)`: marks a write at time `timestamp` as
+  completed.
+
+  All subsequent values of `ReadTs()` will be greater or equal to the given
+  write `timestamp`.
+
+### CATALOG
+
+CATALOG can also be considered a shared service whose methods can be called
+from any component that needs it. And multiple actors can append data to it.
+
+Changes to the catalog are versioned in the `"catalog"` timeline, which is
+tracked separately from other timelines in TIMESTAMP ORACLE. Any successful
+append of updates to the catalog implicitly does an `ApplyWrite("catalog",
+new_upper)` on TIMESTAMP ORACLE.
+
+- `CompareAndAppend(expected_upper, new_upper, updates) -> Result<(),
+  AppendError>`: appends the given updates to the catalog _iff_ the expected
+  upper matches its current upper.
+
+- `SubscribeAt(timestamp) -> Subscribe`: returns a snapshot of the catalog at
+  `timestamp`, followed by an ongoing stream of subsequent updates.
+
+- `SnapshotAt(timestamp)`: returns an in-memory snapshot of the catalog at
+  `timestamp`.
+
+  A snapshot corresponds to our current in-memory representation of the catalog
+  that the Coordinator holds. This can be (and will be, in practice!)
+  implemented using `SubscribeAt` to incrementally keep an in-memory
+  representation of the catalog up to date with changes. However, it is
+  logically a different operation.
+
+### The controllers, aka. STORAGE and COMPUTE
+
+The controllers remain largely as they have been previously designed, but we
+introduce a distinction between deterministic controller command and
+ephemeral/query-like controller commands. See [Controller
+Commands](#controller-commands) above.
+
+The controllers can be seen as consuming a stream of deterministic controller
+commands, in the newly introduced `"catalog"` timeline, and
+`AdvanceCatalogFrontier` is downgrading its frontier.
+
+Ephemeral/query-like commands gain a new `catalog_as_of` parameter that tells
+the controller that it can only process them once its input frontier has
+sufficiently advanced.
```

Comment:
I don't have a good suggestion, but perhaps we can find a better name than `catalog_as_of`? "As of" sounds like an exact time at which we want the peek to happen (like in `SELECT ... AS OF ...`), but this is meant as a lower bound instead.

Reply:
Good point! Maybe we just pick `ts`, and we say these commands can only be processed when the frontier (of catalog updates) is not beyond™️ them anymore? Seems natural. 🤔

Reply:
wait, I think that's also confusing, maybe `catalog_ts`? 🤔

Reply:
Sounds good to me!